python/triton/language/semantic.py

```diff
@@ -1347,6 +1348,24 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, options):
         _0 = builder.get_fp16(0) if out_dtype.is_fp16() else builder.get_fp32(0)
         ret_scalar_ty = out_dtype
 
+    if not allow_tf32 and lhs.dtype.is_fp32() and rhs.dtype.is_fp32() and getattr(builder.options, "supports_tf32"):
+        # Implement 3xTF32 trick
+        # https://github.com/NVIDIA/cutlass/discussions/385
+
+        def fp32_to_tf32(x):
+            return tl.inline_asm_elementwise(asm="""cvt.rna.tf32.f32 $0, $2;\n\tsub.f32 $1, $2, $0;""",
+                                             constraints="=r,=r,r", args=[x], dtype=[tl.float32, tl.float32],
+                                             is_pure=True, pack=1, _builder=builder)
+
+        xb, xs = fp32_to_tf32(lhs)
+        yb, ys = fp32_to_tf32(rhs)
+        del kwargs["lhs"]
+        del kwargs["rhs"]
+        del kwargs["allow_tf32"]
+        kwargs["acc"] = dot(xb, ys, allow_tf32=True, **kwargs)
+        kwargs["acc"] = dot(xs, yb, allow_tf32=True, **kwargs)
+        return dot(xb, yb, allow_tf32=True, **kwargs)
+
```

Comment:
This is cool however my understanding is that the 3xTF32 precision doesn't match exactly fp32 precision which will be a problem for user who want strict precision.

Reply:
Its precision is better than fp32 so that should be fine?

Reply:
I wonder if we should have a separate control for this

Reply:
At least that's fine in PyTorch. Our policy is that `torch.compile` should have the same accuracy or better than eager mode.

Reply:
> Its precision is better than fp32 so that should be fine?

I don't think so, I believe we want it to match

Reply:
> At least that's fine in PyTorch. Our policy is that torch.compile should have the same accuracy or better than eager mode.

We have users that rely on allow_tf32=False to have a precision closer to the CPU

Reply:
I wonder if this is something that can be done on user side? For instance inductor could implement it with a function wrapper?

Reply:
Personally I actually really like this being in the compiler, because it's our speciality to figure out the fastest way to run your code (within whatever correctness constraints we set, of course).  I don't think we want every user to have to re-implement this trick.

Reply:
it could be part of standard but I don't think we should change the precision under the hood

Reply:
No argument against your point that we should not change the precision of the dot without user consent.